### PR TITLE
Fix multi-materials in one skinning model

### LINKED ISSUE #9074: Multi-material GLB model Renders with Incorrect Textures
The GLB model has several materials, all set to the built-in model.material with different textures.
However, at runtime and in the editor, it renders using the texture set for the first material.

To reproduce just run the project 
The example project with glb and blend files is attached [gltf_texture_issue.zip](https://github.com/user-attachments/files/15826205/gltf_texture_issue.zip)

  - Version [1.8.1]

**Platforms (REQUIRED):**
 - Platforms:  Windows
 - OS: Windows 11

image

Here is the setup for the model 
image

**It crashes for** 
  - Version [1.9.0] BETA
  
LOGS: 
```
INFO:DLIB: Log server started on port 53260
INFO:ENGINE: Target listening with name: Busovikov - fe80::2f84:bcca:1766:3e4f - Windows
INFO:ENGINE: Engine service started on port 53261
INFO:GRAPHICS: Installed graphics device 'ADAPTER_FAMILY_OPENGL'
INFO:ENGINE: Defold Engine 1.9.0 (d337da5)
INFO:DLIB: Initialized Remotery (ws://127.0.0.1:17815/rmt)
INFO:ENGINE: Loading data from: build/default
INFO:ENGINE: Initialised sound device 'default'
INFO:CRASH: Successfully wrote Crashdump to file: C:\Users\pavel\AppData\Roaming\Defold/_crash
ERROR:CRASH: CALL STACK:

ERROR:CRASH:  0 0x7FF77D1DBF40 dmCrash::GenerateCallstack D:\a\defold\defold\engine\crash\src\backtrace_win32.cpp:144
ERROR:CRASH:  1 0x7FF77D2D4AFC _seh_filter_exe minkernel\crts\ucrt\src\appcrt\misc\exception_filter.cpp:219
ERROR:CRASH:  2 0x7FF77D319A74 `__scrt_common_main_seh'::`1'::filt$0 D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:304
ERROR:CRASH:  3 0x7FF77D255D5C __C_specific_handler D:\a\_work\1\s\src\vctools\crt\vcruntime\src\eh\riscchandler.cpp:389
ERROR:CRASH:  4 0x7FF9BA6F4F20 __chkstk <unknown>:0
ERROR:CRASH:  5 0x7FF9BA66DDD0 RtlFindCharInUnicodeString <unknown>:0
ERROR:CRASH:  6 0x7FF9BA6F4010 KiUserExceptionDispatcher <unknown>:0
ERROR:CRASH:  7 0x7FF77D0EB830 dmRig::GenerateNormalData D:\a\defold\defold\engine\rig\src\rig.cpp:828
ERROR:CRASH:  8 0x7FF77D0ED560 dmRig::GenerateVertexData D:\a\defold\defold\engine\rig\src\rig.cpp:1249
ERROR:CRASH:  9 0x7FF77CEF41F0 dmGameSystem::RenderBatchWorldVS D:\a\defold\defold\engine\gamesys\src\gamesys\components\comp_model.cpp:965
ERROR:CRASH: 10 0x7FF77CEF53D0 dmGameSystem::RenderListDispatch D:\a\defold\defold\engine\gamesys\src\gamesys\components\comp_model.cpp:1169
ERROR:CRASH: 11 0x7FF77CFDE760 dmRender::DrawRenderList D:\a\defold\defold\engine\render\src\render\render.cpp:932
ERROR:CRASH: 12 0x7FF77D001390 dmRender::ParseCommands D:\a\defold\defold\engine\render\src\render\render_command.cpp:183
ERROR:CRASH: 13 0x7FF77CFEBD10 dmRender::UpdateRenderScriptInstance D:\a\defold\defold\engine\render\src\render\render_script.cpp:3812
ERROR:CRASH: 14 0x7FF77CE20350 dmEngine::StepFrame D:\a\defold\defold\engine\engine\src\engine.cpp:1716
ERROR:CRASH: 15 0x7FF77CE210A0 dmEngineUpdate D:\a\defold\defold\engine\engine\src\engine.cpp:2218
ERROR:CRASH: 16 0x7FF77CE212D0 dmEngine::RunLoop D:\a\defold\defold\engine\engine\src\engine_loop.cpp:83
ERROR:CRASH: 17 0x7FF77CE19430 engine_main D:\a\defold\defold\engine\engine\src\engine_main.cpp:152
ERROR:CRASH: 18 0x7FF77D24A164 __scrt_common_main_seh D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
ERROR:CRASH: 19 0x7FF9B9C52560 BaseThreadInitThunk <unknown>:0
ERROR:CRASH: 20 0x7FF9BA6AAF00 RtlUserThreadStart <unknown>:0
ERROR:CRASH: 

INFO:CRASH: Successfully wrote Crashdump to file: C:\Users\pavel\AppData\Roaming\Defold/_crash
ERROR:CRASH: CALL STACK:

ERROR:CRASH:  0 0x7FF77D0EB830 dmRig::GenerateNormalData D:\a\defold\defold\engine\rig\src\rig.cpp:828
ERROR:CRASH:  1 0x7FF77D0ED560 dmRig::GenerateVertexData D:\a\defold\defold\engine\rig\src\rig.cpp:1249
ERROR:CRASH:  2 0x7FF77CEF41F0 dmGameSystem::RenderBatchWorldVS D:\a\defold\defold\engine\gamesys\src\gamesys\components\comp_model.cpp:965
ERROR:CRASH:  3 0x7FF77CEF53D0 dmGameSystem::RenderListDispatch D:\a\defold\defold\engine\gamesys\src\gamesys\components\comp_model.cpp:1169
ERROR:CRASH:  4 0x7FF77CFDE760 dmRender::DrawRenderList D:\a\defold\defold\engine\render\src\render\render.cpp:932
ERROR:CRASH:  5 0x7FF77D001390 dmRender::ParseCommands D:\a\defold\defold\engine\render\src\render\render_command.cpp:183
ERROR:CRASH:  6 0x7FF77CFEBD10 dmRender::UpdateRenderScriptInstance D:\a\defold\defold\engine\render\src\render\render_script.cpp:3812
ERROR:CRASH:  7 0x7FF77CE20350 dmEngine::StepFrame D:\a\defold\defold\engine\engine\src\engine.cpp:1716
ERROR:CRASH:  8 0x7FF77CE210A0 dmEngineUpdate D:\a\defold\defold\engine\engine\src\engine.cpp:2218
ERROR:CRASH:  9 0x7FF77CE212D0 dmEngine::RunLoop D:\a\defold\defold\engine\engine\src\engine_loop.cpp:83
ERROR:CRASH: 10 0x7FF77CE19430 engine_main D:\a\defold\defold\engine\engine\src\engine_main.cpp:152
ERROR:CRASH: 11 0x7FF77D24A164 __scrt_common_main_seh D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
ERROR:CRASH: 12 0x7FF9B9C52560 BaseThreadInitThunk <unknown>:0
ERROR:CRASH: 13 0x7FF9BA6AAF00 RtlUserThreadStart <unknown>:0
ERROR:CRASH: 

INFO:CRASH: Successfully wrote MiniDump to file: C:\Users\pavel\AppData\Roaming\Defold/_crash.dmp
```
[_crash.dmp](https://github.com/user-attachments/files/15826198/_crash.dmp)

### PR DESCRIPTION
**Pull request by jixingcn**
Fixed an issue where custom vertex attributes are applied only from the first material in a situation where the same material is used across multiple material slots in a model file, but where each sub-material has overridden attributes.

In the .model file, the order of materials must be the same as in the original model file(.glb). (Notice: the defold editor will reorder all materials by name when you edit and save the model.)

### CODE REVIEW

**Comment by jixingcn** [comment]
Reopen this after merging, building, and testing.

**Comment by Jhonnyg** [comment]
Hi! I tried your fix with the original issue, and I don't see that it fixed the issue 🤔 

I think that the problem lies with how instance hashes are calculated, where the overridden textures are not considered. See this PR for more details: https://github.com/defold/defold/pull/9558

**Comment by jixingcn** [comment]
> Hi! I tried your fix with the original issue, and I don't see that it fixed the issue 🤔
> 
> I think that the problem lies with how instance hashes are calculated, where the overridden textures are not considered. See this PR for more details: #9558

OK

> Hi! I tried your fix with the original issue, and I don't see that it fixed the issue 🤔
> 
> I think that the problem lies with how instance hashes are calculated, where the overridden textures are not considered. See this PR for more details: #9558

When reading the code - "engine/gamesys/src/gamesys/components/comp_model.cpp", I found that the material_index is from the first render item when calling `WriteWorldSpaceVertexData`. I get the right result after using the `m_MaterialIndex` of each `render_item`. I am not familiar with the code, but I will check the hints that you mentioned.

I don't test the material with texture and I will do that when I go back home.

**Comment by jixingcn** [comment]
> Hi! I tried your fix with the original issue, and I don't see that it fixed the issue 🤔
> 
> I think that the problem lies with how instance hashes are calculated, where the overridden textures are not considered. See this PR for more details: #9558

Yes, my commit doesn't fix the #9074. But it fixes another issue.
The skinning model is split into several meshes, each with one material in my project. Before this commit, the engine uses only the first material to render all meshes of the skinning model.

|wrong|right|blender|
|-|-|-|
|multimaterial_skinning_model_issue|multimaterial_skinning_model |image|

**Comment by Jhonnyg** [comment]
Could you try your project with https://github.com/defold/defold/pull/9558 to see if that fixes your issue as well?

**Comment by jixingcn** [comment]
> Could you try your project with #9558 to see if that fixes your issue as well?

Sorry, it doesn't fix my issue. I will upload a sample project for my issue later.

**Comment by Jhonnyg** [comment]
Thanks, please do!

**Comment by jixingcn** [comment]
I create a project for testing the multi-material.
[issue-multimaterials.zip](https://github.com/user-attachments/files/17304646/issue-multimaterials.zip)


## Blender:

image


## Test @ v1.9.3:

|VERTEX_SPACE_WORLD|VERTEX_SPACE_LOCAL|
|-|-|
|with skin/animation|no skin/animation|
|image|image|


## Test @ #9558, #9550

|#9558|#9558+#9550|
|-|-|
|image|image|

**For #9558, everything is ok when using the pure-color texture to replace the vertex attribute - `col_tint`.**

**Comment by Jhonnyg** [comment]
Thanks! I'll take a look, Ideally I think I want to know what's going on here before merging any of the PRs here (including mine) :)

**Comment by jixingcn** [comment]
> Thanks! I'll take a look, Ideally I think I want to know what's going on here before merging any of the PRs here (including mine) :)

You're welcome.👍

**Comment by jixingcn** [comment]
By the way, I think the engine also needs to support:

* Get/Set the attribute of each material in a model by the Lua script.

**Comment by Jhonnyg** [comment]
Yes I think so aswell

File: engine/gamesys/src/gamesys/components/comp_model.cpp [hunk 1]
```diff
@@ -1399,6 +1399,7 @@ namespace dmGameSystem
                 dmVMath::Matrix4 normal_matrix    = dmRender::GetNormalMatrix(render_context, world_matrix);
                 bool has_custom_vertex_attributes = vx_decl != world->m_VertexDeclaration;
 
+                uint32_t material_index = render_item->m_MaterialIndex;
                 vb_end = WriteWorldSpaceVertexData(world, render_context, c, render_item, &material_infos_vertex, vertex_stride, material_index, world_matrix, normal_matrix, has_custom_vertex_attributes, vb_end);
```
Comment (Jhonnyg) [inline engine/gamesys/src/gamesys/components/comp_model.cpp:1403]: You were right, it does seem to fix a situation where multiple models are batched, but they have different vertex attributes. Thanks!

**Review by Jhonnyg — APPROVED** [review]

File: engine/gamesys/src/gamesys/components/comp_model.cpp (hunk 1 above)
Comment (jixingcn) [reply engine/gamesys/src/gamesys/components/comp_model.cpp:1403]: You're welcome.

**Comment by jixingcn** [comment]
> I just noticed this issue right now, and to me it feels like we didn't dig into _why_ this happened. The intention was for the batch key tobatch properly. If it doesn't, then there's something wrong somewhere, and we ignored it.

Yes, I agree with you and think there are other potential issues. This is a temporary/quick solution, and I think it doesn't impact others.

**Comment by jixingcn** [comment]
> I just noticed this issue right now, and to me it feels like we didn't dig into _why_ this happened. The intention was for the batch key tobatch properly. If it doesn't, then there's something wrong somewhere, and we ignored it.

Like my other PRs, I don't want to make any big effects when I am a beginner for the impressive **Defold** project.

**Pull request merged**